### PR TITLE
Add `expect_insert` and `unwrap_insert` for associative types.

### PR DESCRIPTION
**Pull request by brennanvincent**
This is occasionally useful.

### CODE REVIEW

**Review by cjubb39 — APPROVED** [review]
sure

File: src/ore/src/collections.rs [hunk 1]
```diff
@@ -65,3 +66,38 @@ where
...
+    }
+}
+
+impl<K, V> AssociativeExt<K, V> for HashMap<K, V>
+where
+    K: Eq + std::hash::Hash,
+{
+    fn expect_insert(&mut self, k: K, v: V, msg: &str) {
+        let old = self.insert(k, v);
+        assert!(old.is_none(), "{}", msg)
```
Comment (cjubb39) [inline src/ore/src/collections.rs:91]: maybe include the old value in the print? I think it'd be pretty informative -- even if it means you have to require that `K: Debug`?

Comment (brennanvincent) [reply src/ore/src/collections.rs:91]: That can't be done in a zero-cost way -- you either have to clone the value when inserting, or do an extra lookup.

Comment (brennanvincent) [reply src/ore/src/collections.rs:91]: (AFAICT)

Comment (cjubb39) [reply src/ore/src/collections.rs:91]: I was suggesting something like `assert!(old.is_none(), "{}: found {:?}", msg, old)` that shouldn't have extra work unless you're about to panic in which case 🤷 

but if you think requiring `K: Debug` makes this less useful, keeping as-is is fine by me

Comment (brennanvincent) [reply src/ore/src/collections.rs:91]: Oh, I see. Yeah, if you could print the key and both the old and new values it'd maybe be useful enough to be outweigh the downside of requiring `Debug`, but I kinda feel like _just_ printing the old value doesn't.

Comment (benesch) [reply src/ore/src/collections.rs:91]: You can make it zero cost with the entry API.

**Pull request merged**